File: Problems/155_minStack.ts
class MinStack {
    //as items get added, the current min is also updated, however if an item is removed, we can simply 'look back in history' for the previous min
    //can also initialise with values since pop(), top() and getMin() are only called with non-empty stacks
    stack: { num: number, currentMin: number }[] = [{ num: Infinity, currentMin: Infinity }];

    constructor() {}

    push(val: number): void {
        const currentMin = Math.min(this.stack[this.stack.length - 1].currentMin, val);
        this.stack.push( { num: val, currentMin: currentMin } );
    }

    pop(): void {
        this.stack.pop();
    }

    top(): number {
        return this.stack[this.stack.length - 1].num;
    }

    getMin(): number {
        return this.stack[this.stack.length - 1].currentMin; //return most recent (lowest) minimum
    }
}

const minStack = new MinStack();
minStack.push(-2);
minStack.push(0);
minStack.push(-3);
console.log(minStack.getMin()); // return -3
minStack.pop();
console.log(minStack.top());    // return 0
console.log(minStack.getMin()); // return -2

File: Problems/155_minStack.py
class MinStack(object):

    stack = []

    def __init__(self):
        self.stack = []

    def push(self, val):
        #as items are pushed onto the stack, we can keep track of the current min and its prior history
        #therefore if an item is removed from the stack (from the top), it will also remove that current min history, thus relegating the min back to its prior value before that item was added
        #this will only change the min if the item removed was the min at the time it was added
        currentMin = self.getMin() if len(self.stack) > 0 else val
        if val < currentMin:
            currentMin = val

        self.stack.append((currentMin, val))

    def pop(self):
        self.stack.pop()


    def top(self):
        return self.stack[len(self.stack) - 1][1]
        

    def getMin(self):
        #we want to know the min of the stack at the current moment
        #use the topmost element (with the most recent history)
        return self.stack[len(self.stack) - 1][0]